### ISSUE TITLE
[Feature Request] Alpha Clusters

### ISSUE THREAD

**Issue by palma21** [Azure/AKS#2058]
Support for sandbox clusters with all feature gates open to allow for testing of new k8s features before they become supported.

**Comment by andyzhangx** [comment]
It's not easy to open all feature gates since some feature gates has hardware or software dependencies, it's better that users could enable alpha feature gate by themselves, just like self config of kubelet parameters.

**Comment by rafaribe** [comment]
Would be awesome to get support for this, looking specifically at the LoadBalancer with Mixed Protocol new feature already mentioned in #2193

**Comment by drzdbd1** [comment]
Interested as well, following.

**Comment by Shaybs** [comment]
Any updates? Ephemeral Containers would be nice, tbh.

**Comment by TilmanLacko-TomTom** [comment]
Looking forward to use `HPAScaleToZero` once this is done..

**Comment by yanrez** [comment]
We are also interested in being able to use feature gates to enable alpha features. In our case it's HostProcess container support on k8s 1.22 . Would be very convenient to have way of enabling alpha features.

**Comment by mehmettaskiner** [comment]
Looking forward to have support for this, we need this to enable mixed protocols in load balancers.

**Comment by philwelz** [comment]
same here, but i would like a solution where i have to specifically enable alpha feature gates instead of have all enabled.

**Comment by annerajb** [comment]
> same here, but i would like a solution where i have to specifically enable alpha feature gates instead of have all enabled.

i agree enabling all of them is way too much compare to selectively enabling individual ones such as `MixedProtocolLBSVC` hpa scale to zero and a few others.

**Comment by TacticalVilius** [comment]
Agree, we also only need the LoadBalancer with mixed protocols feature. Wouldn't want to also enable every other alpha feature.

**Comment by ringerc** [comment]
I couldn't agree more about feature gates being desirable.

For example, I'd like to be able to start development work on supporting ephemeral containers (https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/) before AKS rolls it out as a fully supported option.

**Comment by philwelz** [comment]
> I couldn't agree more about feature gates being desirable.
> 
> For example, I'd like to be able to start development work on supporting ephemeral containers (https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/) before AKS rolls it out as a fully supported option.

You can do this with AKS 1.23 which is public preview at the moment as `ephemeral containers` graduated to beta in 1.23

**Comment by ringerc** [comment]
Excellent, thankyou @philwelz . I incorrectly thought beta features were still gated off by default, but upon rechecking I can see I was mistaken.

**Comment by philwelz** [comment]
> Excellent, thankyou @philwelz . I incorrectly thought beta features were still gated off by default, but upon rechecking I can see I was mistaken.

I already tested with [AKS ](https://twitter.com/philip_welz/status/1493873323427704832?s=20&t=-xhjgcvJsRP_H1eFZSfXcQ)😄

**Comment by doubleyewdee** [comment]
Another bump for allowing enabling of experimental features from Azure ML. We would like MixedProtocolLBService in our case.

**Comment by jamesongithub** [comment]
came here looking for the `StatefulSetAutoDeletePVC` alpha feature released in 1.23

**Comment by yanrez** [comment]
We would like to use StatefulSetStartOrdinal alpha feature gate on 1.26

**Comment by olsenme** [comment]
@rafaribe , @

> Would be awesome to get support for this, looking specifically at the LoadBalancer with Mixed Protocol new feature already mentioned in #2193

Is this still something you are interested in> Would like to learn more about your use case and why you are interested in this.

**Comment by olsenme** [comment]
> Interested as well, following.

@drzdbd1  Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

**Comment by olsenme** [comment]
> Any updates? Ephemeral Containers would be nice, tbh.

@Shaybs Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

**Comment by olsenme** [comment]
> Looking forward to use `HPAScaleToZero` once this is done..

@TilmanLacko-TomTom  Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

**Comment by olsenme** [comment]
> We are also interested in being able to use feature gates to enable alpha features. In our case it's HostProcess container support on k8s 1.22 . Would be very convenient to have way of enabling alpha features.

@yanrez Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

**Comment by olsenme** [comment]
> Looking forward to have support for this, we need this to enable mixed protocols in load balancers.

@mehmettaskiner Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

**Comment by olsenme** [comment]
> same here, but i would like a solution where i have to specifically enable alpha feature gates instead of have all enabled.

@philwelz Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

**Comment by olsenme** [comment]
> Agree, we also only need the LoadBalancer with mixed protocols feature. Wouldn't want to also enable every other alpha feature.

@TacticalVilius Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

**Comment by olsenme** [comment]
> I couldn't agree more about feature gates being desirable.
> 
> For example, I'd like to be able to start development work on supporting ephemeral containers (https://kubernetes.io/docs/concepts/workloads/pods/ephemeral-containers/) before AKS rolls it out as a fully supported option.

@ringerc Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

**Comment by olsenme** [comment]
> came here looking for the `StatefulSetAutoDeletePVC` alpha feature released in 1.23

@jamesongithub Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

**Comment by jamesongithub** [comment]
> > came here looking for the `StatefulSetAutoDeletePVC` alpha feature released in 1.23
> 
> @jamesongithub Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

Please google `StatefulSetAutoDeletePVC`

**Comment by doubleyewdee** [comment]
back again, we got mixed protocol LB, but now we're interested in HPAContainerMetrics. this issue is really old at this point, it would be nice to either get some commentary from AKS about generally offering alpha-gated k8s features or simply be told "no" and have this issue closed.

thanks!

**Comment by olsenme** [comment]
> back again, we got mixed protocol LB, but now we're interested in HPAContainerMetrics. this issue is really old at this point, it would be nice to either get some commentary from AKS about generally offering alpha-gated k8s features or simply be told "no" and have this issue closed.
> 
> thanks!

What are your thoughts if we were to support only latest stable version of AKS for alpha clusters? Would this meet your needs?

**Comment by dhirschfeld** [comment]
> What are your thoughts if we were to support only latest stable version of AKS for alpha clusters? Would this meet your needs?

If you're wanting to try out *alpha* features it seems you'd more likely want to test them on the latest preview cluster.

**Comment by bok11** [comment]
> > Looking forward to use `HPAScaleToZero` once this is done..
> 
> @TilmanLacko-TomTom Is this still something you are interested in? Would like to learn more about your use case and why you are interested in this.

This is something I am interested in as well.
My use case is this:

My microservices work in a choreography architecture, so I only need them to be on if there are work for them to do.
We use devspace to develop, so we have many environments with idle pods running.

Additionally, one of our workloads require 200gb of memory, so if 5 developers each have their own environment they reserve 1 Tb of memory just to have their helm chart deployed, but without having any actual workload.

**Comment by olsenme** [comment]
Anyone else thoughts on the proposal to support only on latest stable version?

**Comment by doubleyewdee** [comment]
> > back again, we got mixed protocol LB, but now we're interested in HPAContainerMetrics. this issue is really old at this point, it would be nice to either get some commentary from AKS about generally offering alpha-gated k8s features or simply be told "no" and have this issue closed.
> > thanks!
> 
> What are your thoughts if we were to support only latest stable version of AKS for alpha clusters? Would this meet your needs?

100% would meet our needs yes. we're happy to upgrade/roll-forward any time we need these things, also happy to flip a specific bit at cluster provisioning time and even totally rebuild our clusters to get features turned on.

**Comment by bok11** [comment]
> Anyone else thoughts on the proposal to support only on latest stable version?

This would be fine for me, it shouldn't be an issue to keep our cluster updates to the latest version.

**Comment by pohly** [comment]
@olsenme: this was added to the backlog on April 13 and then removed again completely on June 5. Does that mean that it's no longer planned?

I'd like to add one more aspect: besides enabling feature gates, sometimes a feature also needs API groups enabled in the apiserver. Example (my current use case): https://kubernetes.io/docs/concepts/scheduling-eviction/dynamic-resource-allocation/#enabling-dynamic-resource-allocation

**Comment by dtzar** [comment]
CAPZ will support creating and joining alpha/beta worker nodes to AKS control plane.  For more information see [this PR ](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/4052) and [recorded demo](https://youtu.be/braHIOyJt5o) from the community meeting.
If you're interested to use this, functionality please feel free to ping me or [join CAPZ office hours or slack](https://github.com/kubernetes-sigs/cluster-api-provider-azure#office-hours)

**Comment by olsenme** [comment]
What current alpha features are you looking to take advantage of? What would be the advantages of having them in the alpha state vs beta stage?

**Comment by olsenme** [comment]
> Looking forward to use `HPAScaleToZero` once this is done..

What is your use case?

**Comment by AndreiDoroftei** [comment]
Hello @olsenme! In addition to this topic, I would love to have the InPlacePodVerticalScaling feature gate enabled ASAP. We have many Spring applications that would use this functionality. They use a lot of CPU during startup, and afterwards, there is almost no usage. If feature gates can't be enabled manually, could you please provide me with an estimated date for enabling this specific feature gate? Thank you in advance!

**Comment by colinmixonn** [comment]
**Update: Reopening Alpha Clusters Discussion**
**Status:** Backlog

Reopening this discussion to share progress on AKS Alpha Clusters. Alpha Clusters will enable ALL Alpha and Beta feature-gates, and Alpha and Beta APIs by DEFAULT
